tools/autograd/gen_variable_type.py

```diff
@@ -1918,20 +1942,43 @@ def emit_fw_derivatives() -> List[str]:
                     f"std::vector<{opt_res_grad_type}> {'_'.join(res)}_new_fw_grad_opts"
                     "(self.size(), c10::nullopt);"
                 )
-                inplace_foreach_forward_grad_formula = derivative.formula
-                for _foreach_arg, _ref_arg in inplace_foreacharg2refarg.items():
-                    # note(crcrpar): Massage only Scalar and ArrayRef<Scalar> here.
-                    if not (
-                        is_tensor_type(_foreach_arg.type)
-                        or is_tensor_list_type(_foreach_arg.type)
-                    ):
-                        pattern = _foreach_arg.name
-                        if isinstance(_foreach_arg.type, ListType):
-                            pattern += "[i]"
-                        inplace_foreach_forward_grad_formula = (
-                            inplace_foreach_forward_grad_formula.replace(
-                                _ref_arg.name, pattern
+                foreach_forward_grad_formula = derivative.formula
+                _foreach_arg: Union[Argument, DifferentiableInput]
+                if inplace:
+                    for _foreach_arg, _ref_arg in inplace_foreacharg2refarg.items():
+                        # note(crcrpar): Massage only Scalar and ArrayRef<Scalar> here.
+                        if not (
+                            is_tensor_type(_foreach_arg.type)
+                            or is_tensor_list_type(_foreach_arg.type)
+                        ):
+                            pattern = _foreach_arg.name
+                            if isinstance(_foreach_arg.type, ListType):
+                                pattern += "[i]"
+                            foreach_forward_grad_formula = (
+                                foreach_forward_grad_formula.replace(
+                                    _ref_arg.name, pattern
+                                )
+                            )
+                else:
+                    for _foreach_arg in differentiable_inputs:
```

Comment:
If we didn't update `gen_foreach_derivativeinfo` to replace `_ref_arg` with `foreach_arg` when we created the ForwardDerivative, could we avoid the extra if-else checking whether we are inplace or not? 

Or, we could also update the inplace case to create a new DifferentiabilityInfo (in which the grad_fn/node is still the same)? (you might have advocated for some version of this idea in https://github.com/pytorch/pytorch/pull/96405)

Reply:
this part is turning out to be vain. I'll remove this for-loop with if inside of it once I confirmed it